modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java

```diff
@@ -1809,21 +1809,31 @@ private CompletableFuture<Void> handleChangePendingAssignmentEvent(
 
         Assignments nonStableNodeAssignmentsFinal = nonStableNodeAssignments;
 
+        int partitionId = replicaGrpId.partitionId();
+
         if (shouldStartLocalGroupNode) {
+            var singlePartitionIdSet = PartitionSet.of(partitionId);
+
             localServicesStartFuture = localPartitionsVv.get(revision)
                     // TODO https://issues.apache.org/jira/browse/IGNITE-20957 Revisit this code
-                    .thenApply(unused -> localPartsByTableId.get(tableId).copy())
-                    .thenComposeAsync(partitionSet -> inBusyLock(busyLock,
-                            () -> getOrCreatePartitionStorages(tbl, partitionSet)
-                    ), ioExecutor)
-                    .thenComposeAsync(unused -> inBusyLock(busyLock, () -> startPartitionAndStartClient(
-                            tbl,
-                            replicaGrpId.partitionId(),
-                            pendingAssignments,
-                            nonStableNodeAssignmentsFinal,
-                            zoneId,
-                            isRecovery
-                    )), ioExecutor);
+                    .thenComposeAsync(
+                            partitionSet -> inBusyLock(busyLock, () -> getOrCreatePartitionStorages(tbl, singlePartitionIdSet)),
+                            ioExecutor
+                    )
+                    .thenComposeAsync(unused -> inBusyLock(busyLock, () -> {
+                        if (!isRecovery) {
+                            registerIndexesToTableOnNodeRecoveryOrRebalance(tbl, catalogService, singlePartitionIdSet, tbl.schemaView());
+                        }
```

Comment:
I don't understand what's happening. Please comment.
This code will probably only be executed during a rebalance. By that time, table should already be created and indexes should already be registered. Am I missing something?

Maybe this method should be renamed, because "register ... to table" is misleading. What you really do is you do something with individual partitions. Right?